3rdParty/fuerte/src/H2Connection.h

```diff
@@ -0,0 +1,149 @@
+  //  - a messageid
+  //  - the buffer to be send
+  // this item is then moved to the request queue
+  // and a write action is triggerd when there is
+  // no other write in progress
+  void sendRequest(std::unique_ptr<Request>, RequestCallback) override;
+
+  // Return the number of unfinished requests.
+  std::size_t requestsLeft() const override;
+
+ protected:
+  void finishConnect() override;
+
+  // Thread-Safe: activate the writer loop (if off and items are queud)
+  void startWriting() override;
+
+  // called by the async_read handler (called from IO thread)
+  void asyncReadCallback(asio_ns::error_code const&) override;
+
+  /// abort ongoing / unfinished requests (locally)
+  void abortOngoingRequests(const fuerte::Error) override;
+
+  /// abort all requests lingering in the queue
+  void drainQueue(const fuerte::Error) override;
+
+ private:
+  static int on_begin_headers(nghttp2_session* session,
+                              const nghttp2_frame* frame, void* user_data);
+  static int on_header(nghttp2_session* session, const nghttp2_frame* frame,
+                       const uint8_t* name, size_t namelen,
+                       const uint8_t* value, size_t valuelen, uint8_t flags,
+                       void* user_data);
+  static int on_frame_recv(nghttp2_session* session, const nghttp2_frame* frame,
+                           void* user_data);
+  static int on_data_chunk_recv(nghttp2_session* session, uint8_t flags,
+                                int32_t stream_id, const uint8_t* data,
+                                size_t len, void* user_data);
+  static int on_stream_close(nghttp2_session* session, int32_t stream_id,
+                             uint32_t error_code, void* user_data);
+  static int on_frame_not_send(nghttp2_session* session,
+                               const nghttp2_frame* frame, int lib_error_code,
+                               void* user_data);
+
+ private:
+  // ongoing Http2 stream
+  struct Stream {
+    velocypack::Buffer<uint8_t> data;
+
+    RequestCallback callback;
+    std::unique_ptr<arangodb::fuerte::v1::Request> request;
+    std::unique_ptr<arangodb::fuerte::v1::Response> response;
+    size_t responseOffset = 0;
+    /// point in time when the message expires
+    std::chrono::steady_clock::time_point expires;
+
+    inline void invokeOnError(Error e) {
+      callback(e, std::move(request), nullptr);
+    }
+  };
+
+  void initNgHttp2Session();
+  
+  void readSwitchingProtocolsResponse();
+
+  // adjust the timeouts (only call from IO-Thread)
+  void setTimeout();
+
+  /// should close connection
+  bool shouldStop() const;
+
+  // queue the response onto the session, call only on IO thread
+  void queueHttp2Requests();
+
+  /// actually perform writing
+  void doWrite();
+
+  Stream* findStream(int32_t sid) const;
+
+ private:
+  static constexpr size_t kOutBufferLen = 32 * 1024 * 1024;
+  std::array<uint8_t, kOutBufferLen> _outbuffer;
```

Comment:
This generates a 32 MB buffer per HTTP/2 connection. This seems overly much.
Can't we allocate this buffer on the heap if needed?